pingouin/contingency.py

```diff
@@ -345,6 +345,188 @@ def chi2_mcnemar(data, x, y, correction=True):
+
+    >>> observed
+    target      0     1
+    sex
+    0        24.5  71.5
+    1       113.5  93.5
+
+    The proportion is lower on the class 0 and higher on the class 1. The
+    tests should be sensitive to this difference.
+
+    >>> stats.round(3)
+              alternative  odds ratio    pval
+    0           two-sided       0.272     0.0
+    1                less       0.272     0.0
+    2             greater       0.272     1.0
+
+    The p-value is very small when the alternative is two-sided and less. 
+    suggests that there is a significant relationship between gender and 
+    heart disease, and that women are less likely to suffer from heart disease
+    than men.
+    """
+    # Python code inspired by SciPy's fisher_exact,barnard_exact,boschloo_exact
+    assert isinstance(data, pd.DataFrame), "data must be a pandas DataFrame."
+    assert isinstance(x, (str, int)), "x must be a string or int."
+    assert isinstance(y, (str, int)), "y must be a string or int."
+    assert all(col in data.columns for col in (x, y)), "columns are not in dataframe."
+    assert isinstance(method, str), "method must be a string and must be one of 'fisher', 'barnard', 'boschloo'."
+
+    observed = pd.crosstab(data[x], data[y])
+
+    if observed.size == 0:
+        raise ValueError("No data; observed has size 0.")
+
+    expected = pd.DataFrame(expected_freq(observed), index=observed.index, columns=observed.columns)
+    
+    stats = []
+    alternatives = [
+        "two-sided",
+        "less",
+        "greater",
+    ]
+    
+    if method == 'fisher':
+        for alternative in alternatives:
+            odds, pval = fisher_exact(table = observed, alternative = alternative)
+            stats.append(
+            {
+                "alternative": alternative,
+                "odds ratio": odds,
+                "pval": pval,
+            }
+        )
+            
+        stats = pd.DataFrame(stats)[["alternative", "odds ratio", "pval"]]
+
+    elif method == 'barnard':
+        for alternative in alternatives:
+            res = barnard_exact(table = observed, alternative = alternative, **kwargs)
+            odds, pval = (res.statistic, res.pvalue)
+            stats.append(
+            {
+                "alternative": alternative,
+                "odds ratio": odds,
+                "pval": pval,
+            }
+        )
+            
+        stats = pd.DataFrame(stats)[["alternative", "odds ratio", "pval"]]
+    
+    elif method == 'boschloo':
+        for alternative in alternatives:
+            res = boschloo_exact(table = observed, alternative = alternative, **kwargs)
+            odds, pval = (res.statistic, res.pvalue)
+            stats.append(
+            {
+                "alternative": alternative,
+                "odds ratio": odds,
+                "pval": pval,
+            }
+        )
+            
+        stats = pd.DataFrame(stats)[["alternative", "odds ratio", "pval"]]
```

Comment:
You can easily reduce code redundancy:
```suggestion
    if method == "fisher":
        func = fisher_exact
    elif method == "barnard":
        func = barnard_exact
    elif method == "boschloo":
        func = boschloo_exact
    else:
        raise ValueError(f"Unsupported method {method}")

    for alternative in alternatives:
        res = func(table=observed, alternative=alternative, **kwargs)
        stats.append(
            {
                "alternative": alternative,
                "odds ratio": res.statistic,
                "pval": res.pvalue,
            }
        )

    stats = pd.DataFrame(stats)[["alternative", "odds ratio", "pval"]]
```
A suggestion is to transform `method` argument from a string into any function with a defined signature.